src/web/src/topology_input_power_chain.ecpp

```diff
@@ -0,0 +1,165 @@
+ * \brief  process topology input powerchain request via 'asset-agent' TOPOLOGY mailbox
+ */
+ #><%pre>
+#include <fty_common.h>
+#include <fty_common_macros.h>
+#include <fty_common_rest_helpers.h>
+#include <fty_common_db.h>
+#include <fty_common_mlm_tntmlm.h>
+
+// set S with MSG popped frame (S unchanged if NULL frame)
+static void zmsg_pop_s (zmsg_t *msg, std::string & s)
+{
+    char *aux = msg ? zmsg_popstr (msg) : NULL;
+    if (aux) { s = aux; zstr_free (&aux); }
+}
+</%pre>
+<%request scope="global">
+    UserInfo user;
+    bool database_ready;
+</%request>
+<%cpp>
+{
+    // verify server is ready
+    if (!database_ready) {
+        log_debug ("Database is not ready yet.");
+        http_die ("internal-error", "Database is not ready yet, please try again after a while.");
+    }
+
+    // check user permissions
+    static const std::map <BiosProfile, std::string> PERMISSIONS = {
+            {BiosProfile::Admin,         "R"},
+            {BiosProfile::Dashboard,     "R"}
+    };
+    CHECK_USER_PERMISSIONS_OR_DIE (PERMISSIONS);
+
+    //ftylog_setVeboseMode(ftylog_getInstance());
+    log_trace ("in %s", request.getUrl().c_str ());
+
+    const char *ADDRESS = AGENT_FTY_ASSET; // "asset-agent" 42ty/fty-asset
+    const char *SUBJECT = "TOPOLOGY";
+    const char *COMMAND = "INPUT_POWERCHAIN";
+    const int TIMEOUT = 5; // seconds
+
+    // get param, id of datacenter retrieved from url
+    std::string asset_id = request.getArg ("id");
+    if (asset_id.empty ()) {
+        std::string expected = TRANSLATE_ME("existing asset name");
+        http_die ("request-param-bad", "dc_id", asset_id.c_str (), expected.c_str ());
+    }
+
+    // db checks
+    {
+        // asset_id valid?
+        if (!persist::is_ok_name (asset_id.c_str ()) ) {
+            std::string expected = TRANSLATE_ME("valid asset name");
+            http_die ("request-param-bad", "dc_id", asset_id.c_str (), expected.c_str ());
+        }
+        // asset_id exist?
+        int64_t rv = DBAssets::name_to_asset_id (asset_id);
+        if (rv == -1) {
+            http_die ("element-not-found", asset_id.c_str ());
+        }
+        if (rv == -2) {
+            std::string err = TRANSLATE_ME("Connection to database failed.");
+            http_die ("internal-error", err.c_str ());
+        }
+    }
+
+    log_trace ("%s, asset_id: '%s'",
+        request.getUrl().c_str (), asset_id.c_str ());
+
+    // connect to mlm client
+    MlmClientPool::Ptr client = mlm_pool.get ();
+    if (!client.getPointer ()) {
+        log_error ("mlm_pool.get () failed");
+        std::string err = TRANSLATE_ME("Connection to mlm client failed.");
+        http_die ("internal-error", err.c_str());
+    }
+
+    // set/send req, recv response
+    zmsg_t *req = zmsg_new ();
+    if (!req) {
+        log_error ("zmsg_new () failed");
+        std::string err = TRANSLATE_ME("Memory allocation failed.");
+        http_die ("internal-error", err.c_str());
```

Comment:
Inline string construction:
```c++
http_die ("internal-error", TRANSLATE_ME("Memory allocation failed.").c_str());
```

Reply:
http_die is a 'awfull-but-necessary' macro, as TRANSLATE_ME(). How many times TRANSLATE_ME would be instanciated by http_die()? Nobody knows... I prefer the current 'var' schema, personally. ;)